### PR TITLE
Adds support for InstanceProviders for on demand newly constructed object instances

### PR DESCRIPTION
**Pull request by nzack**
Adds functionality for instance providers that construct a newly created instance each time they're called.  When the object is accessed from the container it receives an InstanceProvider object that contains one method, `instance`, that will return a newly created, fully constructed instance.

```
bottle.instanceProvider('beer', (container) => { return new IPA(container.hops, container.barley); });

let beerInstanceProvider = bottle.container.beer;

function beerMe() {
  return beerInstanceProvider.instance();
}

beerMe();  // First beer
beerMe();  // Second beer
```

This is similar to Google Guice's InjectingProviders functionality. https://github.com/google/guice/wiki/InjectingProviders

My specific use-case for this is creating objects that only need to live during the lifetime of a request and are specific to the request coming in.  These objects store request state and as soon as the request is done the objects can be disposed of.

FYI: When I run `grunt test` everything freezes on the Jasmine tests...
```
Running "jasmine:tests" (jasmine) task
Testing jasmine specs via PhantomJS
```
Then it just sits there.. I've tried Node 0.10, 4.4, 5.10, and 6.2... no luck on any of them.... else I would've added real tests.  Any thoughts on how I could get that set up?

This library is awesome btw

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -332,6 +332,32 @@ Param    | Type     | Details
...
 
+#### instanceProvider(name, Factory)
+
+Used to register a service instance provider that will return a newly created, fully configured
+instance when called.
+
+Param       | Type       | Details
+:-----------|:-----------|:--------
+**name**    | *String*   | The name of the service.  Must be unique to each Bottle instance.
+**Factory** | *Function* | A function that should return the a fully configured service object.  This factory method will be called whenever a new instance is created.  Gets passed an instance of the container to allow dependency injection when creating a new instance of the service.
```
Comment (young-steveo) [inline README.md]: Small typo.  Should read:

A function that should return a fully configured service object.  This factory function will be called whenever a new instance is created.  Gets passed an instance of the container to allow dependency injection when creating a new instance of the service.

File: dist/bottle.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ;(function(undefined) {
     'use strict';
     /**
-     * BottleJS v1.3.0 - 2016-04-29
+     * BottleJS v1.3.0 - 2016-07-06
```
Comment (young-steveo) [inline dist/bottle.js]: @nzack Can you remove the dist generated files from your PR?  I'll generate them when a new bottle version is released.

File: src/Bottle/instance-provider.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/**
+ * Register an instance provider inside a generic provider.
+ *
+ * @param {String} name - The name of the service
+ * @param {Function} Factory - The factory function, matches the signature required for the
+ * `factory` method
+ * @return Bottle
+ */
+var instanceProvider = function instanceProvider(name, Factory) {
+  return provider.call(this, name, function GenericProvider() {
```
Comment (young-steveo) [inline src/Bottle/instance-provider.js]: @nzack I think you can simplify this by doing something like the following:

```js
var instanceProvider = function instanceProvider(name, Factory) {
  return factory.call(this, name, function GenericInstanceFactory(container) {
      return {
        instance : Factory.bind(null, container)
      };
  });
};
```

Then you can drop the new function constructer below.

**Comment by young-steveo** [comment]
Thanks for the PR!  This is an often requested feature that I haven't supported because I couldn't think of a good way to do it.  However, I like this concept, so thanks!

I'm not sure why you're having trouble running the tests.  They execute fine on my machine, and also on the Travis CI server.

Obviously we'll want to have good unit test coverage.  However, if you're still having trouble with it, make the changes that I've suggested so far, and I'll write tests for it after it's merged.

File: src/Bottle/instance-provider.js (hunk 3 above)
Comment (nzack) [inline src/Bottle/instance-provider.js]: good call, making the updates now

**Comment by nzack** [comment]
Hey no problem.  Had this need myself and remembered how it was done in Google Guice and figured I'd add it in.

I'm still not able to run the tests... still not sure why..

Here's the output from `grunt test --verbose`
```
Running "jasmine" task

Running "jasmine:tests" (jasmine) task
Verifying property jasmine.tests exists in config...OK
Files: dist/bottle.js
Options: version="2.0.1", timeout=10000, styles=[], specs="test/spec/*.spec.js", helpers=[], vendor=[], outfile="_SpecRunner.html", host="", template="/Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/jasmine/templates/DefaultRunner.tmpl", templateOptions={}, junit={}, ignoreEmpty=false, display="full", summary=false
Reading /Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/lib/../jasmine/reporters/PhantomReporter.js...OK
Writing .grunt/grunt-contrib-jasmine/reporter.js...OK
Reading /Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/lib/../../vendor/jasmine-2.0.1/jasmine.css...OK
Writing .grunt/grunt-contrib-jasmine/jasmine.css...OK
Reading /Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/lib/../../vendor/jasmine-2.0.1/jasmine.js...OK
Writing .grunt/grunt-contrib-jasmine/jasmine.js...OK
Reading /Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/lib/../../vendor/jasmine-2.0.1/jasmine-html.js...OK
Writing .grunt/grunt-contrib-jasmine/jasmine-html.js...OK
Reading /Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/lib/../../vendor/jasmine-2.0.1/boot.js...OK
Writing .grunt/grunt-contrib-jasmine/boot.js...OK
Reading /Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/lib/../../vendor/jasmine-2.0.1/jasmine_favicon.png...OK
Writing .grunt/grunt-contrib-jasmine/jasmine_favicon.png...OK
Reading /Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/lib/../../node_modules/es5-shim/es5-shim.js...OK
Writing .grunt/grunt-contrib-jasmine/es5-shim.js...OK
Reading /Users/znelson/git/dev/temp/bottlejs/node_modules/grunt-contrib-jasmine/tasks/jasmine/templates/DefaultRunner.tmpl...OK
Processing source...OK
Writing _SpecRunner.html...OK

Testing jasmine specs via phantom

```

I'm on vacation this week, not sure how much time i'll have to investigate, but will try more when i'm back.

**Comment by young-steveo** [comment]
@nzack If you can remove your current test file, I'll merge and get some tests written for it.

**Comment by young-steveo** [comment]
@nzack If you are back from your vacation, just remove the test file from this PR and I'll get it merged.

**Comment by nzack** [comment]
k, all done.  sorry been swamped since i got back from vacation

**Comment by nzack** [comment]
@young-steveo Hey, just checking in on this.  Did you need anything else from me?

**Comment by young-steveo** [comment]
@nzack Sorry, somehow I thought I already merged.  🐳

**Comment by young-steveo** [comment]
@nzack I'll add tests tonight and if everything looks good, I'll bump the version and distros.

**Pull request merged**